### PR TITLE
Classifier: Consolidate store subject observers

### LINKED ISSUE #1521: Classifier: Last workflow step is still selected when a new classification starts
**Package**
lib-classifier

**Describe the bug**
On starting a new classification for the second, and subsequent, subject in the queue, the first annotation that is added to the new classification is the annotation for the last task in the workflow.

Here we log calls to `classifications.addAnnotation(task, value)` on classification complete. Task `T3`, the last task, is added to the new classification before `T0`, the first task.
Screenshot of the console log, showing annotations being added to a classification out of order.

When the classification is submitted, the order of the annotations array is `[ T3, T0, T1, T2]`, not `[T0, T1, T2, T3]`.

**To Reproduce**
Classify more than one subject on project 1844, which has four steps.
http://localhost:8080/?project=1844

### PR DESCRIPTION
**Pull request by eatyourgreens**
Consolidate the subject observers from the feedback, workflow steps and classifications stores into one function in the root store. When the subject changes, reset the workflow steps, create a new classification, then reset subject feedback. I think this will fix #1521 by making sure the first task is selected before the classification is created.

This PR also fixes a couple of infinite loops that showed up in the Tutorial Store tests (infinitely requesting new subjects) and the UPP Store tests (infinitely requesting a new workflow.) 

I've also moved the Feedback `onAction` listener up to the root in order to mute warnings from that store. Classification and Mark `onAction` listeners are replaced with `beforeDestroy` actions to remove those warnings too. Otherwise, I was finding the console so noisy that failing tests were impossible to debug.

Package:
lib-classifier

Closes #1521.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Testing this is throwing up a bunch of infinite loops in the mock Root Stores used in the UPP tests. I'm looking into that to see if I can figure out what's going on.

**Comment by eatyourgreens** [comment]
UPP is in a loop because it constantly requests a workflow but receives nothing.

**Comment by eatyourgreens** [comment]
Tests are fixed now so I think this is ready for review.

I've also moved the feedback onAction listener up to the root in order to mute some of the onAction warnings.

**Comment by shaunanoordin** [comment]
## Quick Review

Functionally, this PR looks legit - the changes to the code correctly ensures that annotations are made in the correct order, _for the 2nd Classification onwards._ (The bug does not affect the 1st Classification submitted. See #1521 for deets.)

e.g. for a Workflow with 3 Tasks, the annotation values submitted to Panoptes _on this branch_ are in the form of `[ { task: 'T0', ... }, { task: 'T1', ... }, { task: 'T2', ... } ]`, which is expected.

This is instead of the form of `[ { task: 'T2', ... }, { task: 'T0', ... }, { task: 'T1', ... } ]`, which is what's going on in `master` (6a913857ace00b1f1abb83678713d66397395dac)

This PR gets a 👍 from a functional point of view, though I'm going to take my time to grok the actual code changes.

**Comment by eatyourgreens** [comment]
I'm going to try changing this to run `onSubjectReady` too. I think that would make it more clear that we're resetting the workflow etc. after a new subject loads.

**Comment by eatyourgreens** [comment]
I've updated this to reset the workflow, create a new classification and reset feedback on subject ready. I've also added tests that check the store state after `onSubjectReady` runs.

Mocking the root store for the tests was a lot harder than I expected. There was a lot of reading through store code to figure what needed to be set and which actions should be called to set up the initial store state. It turns out that the root store can't be created directly from a snapshot. You must use `ResourceStore.setResource` and `ResourceStore.setActive` to set up workflows and subjects. That could be improved, maybe as part of #1554.

**Comment by eatyourgreens** [comment]
Is there a project where we can test this with feedback enabled?

**Comment by eatyourgreens** [comment]
The classification store tests for 'with valid feedback > classification metadata' are throwing MST warnings.
```
Error: [mobx-state-tree] You are trying to read or write to an object that is no
longer part of a state tree. (Object type: 'ClassificationMetadata',
Path upon death: '/classifications/resources/ck7lzu84s0066lktw7beca3gn/metadata',
Subpath: 'feedback', Action: '').
Either detach nodes first, or don't use objects after removing / replacing
them in the tree.
```

I'm not seeing any warnings in the browser so I'm going to put that down to something being not quite right in the mocks for those tests. I can't immediately see what it is.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js [hunk 1]
```diff
@@ -245,7 +255,7 @@ describe('Model > ClassificationStore', function () {
       // Why is this test here?
       // The observer is in the feedback store
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/ClassificationStore.spec.js]: I think I’ve moved this observer in this PR.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js [hunk 2]
```diff
@@ -111,8 +117,10 @@ describe('Model > ClassificationStore', function () {
         workflowSteps: {},
         userProjectPreferences: {}
       })
-      rootStore.subjects.setResource(subjectsStub[0])
-      rootStore.subjects.setActive(subjectsStub[0].id)
+      rootStore.subjectViewer.onSubjectReady()
```
Comment (srallen) [inline packages/lib-classifier/src/store/ClassificationStore.spec.js]: Same comment about calling `onSubjectReady` here.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js (hunk 1 above)
Comment (srallen) [reply packages/lib-classifier/src/store/ClassificationStore.spec.js]: Non-blocking comment: Yep. Will these specs move to the `RootStore` spec file? At the very least the comment about the observer can be removed.

File: packages/lib-classifier/src/store/RootStore.js [hunk 3]
```diff
@@ -33,6 +34,49 @@ const RootStore = types
...
+        })
+      }, { name: 'Root Store Classification Observer autorun' })
+      addDisposer(self, classificationDisposer)
+    }
+
+    function createSubjectObserver () {
+      const subjectDisposer = autorun(() => {
+        onAction(self, (call) => {
+          if (call.name === 'onSubjectReady') {
+            onSubjectReady()
```
Comment (srallen) [inline packages/lib-classifier/src/store/RootStore.js]: Nice. This is easier to read and clearer on what is triggering the reset.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js [hunk 4]
```diff
@@ -66,8 +70,10 @@ describe('Model > ClassificationStore', function () {
...
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/ClassificationStore.spec.js]: This broke the tests when I ran it inside `setupStores`, after the active subject is set there, so I moved it out. I'll take another look.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/ClassificationStore.spec.js]: This should move up to the root store tests. So much is being stubbed out here that I find it hard to follow what's really being tested. When these tests are moved, I'd suggest rewriting them to remove the stubs and test expected feedback state after a classification completes instead.

File: packages/lib-classifier/src/store/ClassificationStore.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/ClassificationStore.spec.js]: I remember. This is here to work around a problem with the feedback tests. I've rewritten the setup function, and added a comment about the workaround.

**Comment by eatyourgreens** [comment]
I think the drawing task crash might be happening when the subject renders `onReady` but before the new classification has been created. So it's trying to render annotations from the submitted classification.

**Comment by eatyourgreens** [comment]
>I think this can be fixed by checking for the SubjectViewerStore's loading state in InteractionLayerContainer to be success before rendering.

That worked, thanks. I changed `enableInteractionLayer` to check the loading state too.

File: packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js [hunk 5]
```diff
@@ -41,21 +39,21 @@ class Tasks extends React.Component {
   }
 
   [asyncStates.success] () {
-    const { addAnnotation, classification, isThereTaskHelp, subjectReadyState, step, tasks, theme } = this.props
+    const { classification, isThereTaskHelp, subjectReadyState, step, theme } = this.props
     const ready = subjectReadyState === asyncStates.success
-    if (classification && tasks.length > 0) {
+    if (classification && step.tasks.length > 0) {
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js]: [minor] there's no instance where `step` can be undefined, right?

File: packages/lib-classifier/src/store/RootStore.js [hunk 6]
```diff
@@ -33,6 +34,49 @@ const RootStore = types
   })
 
   .actions(self => {
+    // Private methods
+    function onSubjectReady () {
+      const { classifications, feedback, projects, subjects, workflows, workflowSteps } = self
+      const subject = tryReference(() => subjects?.active)
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/store/RootStore.js]: [aside] Wait, we have `?.` existence checkers?? This is cool, I didn't realise we had this - where's it coming from? Babel? ES2020?

File: packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js (hunk 5 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js]: I don't think so. I _think_ the model always has a `tasks` property, which defaults to `[]`.

**Comment by eatyourgreens** [comment]
>(If I'm understanding the cause of issue #1521 is that there's some sort of race condition where the indvidualistic sub-stores are basically performing their cleanup operations regardless of what the other sub-stores are doing, so e.g. annotations stores are resetting before the task stores, or some sort of similar madness.)

I don't fully understand it, but I think this is the case. We don't have any control over the order in which the individual subject observers run in the workflow steps, classifications and feedback models. My idea was to consolidate them into a single function and run them in a prescribed order.

File: packages/lib-classifier/src/store/RootStore.js (hunk 6 above)
Comment (srallen) [reply packages/lib-classifier/src/store/RootStore.js]: [Optional chaining operator](https://github.com/tc39/proposal-optional-chaining)! It's stage 4 now, so it's part of the default babel support and should be coming to the browser soon!
